runtime/parachains/src/router/ump.rs

```diff
@@ -0,0 +1,717 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::{Trait, Module, Store};
+use crate::configuration::{self, HostConfiguration};
+use sp_std::prelude::*;
+use sp_std::collections::{btree_map::BTreeMap, vec_deque::VecDeque};
+use frame_support::{StorageMap, StorageValue, weights::Weight, traits::Get};
+use primitives::v1::{Id as ParaId, UpwardMessage};
+
+/// All upward messages coming from parachains will be funneled into an implementation of this trait.
+///
+/// The message is opaque from the perspective of UMP. The message size can range from 0 to
+/// `config.max_upward_message_size`.
+///
+/// It's up to the implementation of this trait to decide what to do with a message as long as it
+/// returns the amount of weight consumed in the process of handling. Ignoring a message is a valid
+/// strategy.
+///
+/// There are no guarantees on how much time it takes for the message sent by a candidate to end up
+/// in the sink after the candidate was enacted. That typically depends on the UMP traffic, the sizes
+/// of upward messages and the configuration of UMP.
+///
+/// It is possible that by the time the message is sank the origin parachain was offboarded. It is
+/// up to the implementer to check that if it cares.
+pub trait UmpSink {
+	/// Process an incoming upward message and return the amount of weight it consumed.
+	///
+	/// See the trait docs for more details.
+	fn process_upward_message(origin: ParaId, msg: Vec<u8>) -> Weight;
+}
+
+/// An implementation of a sink that just swallows the message without consuming any weight.
+impl UmpSink for () {
+	fn process_upward_message(_: ParaId, _: Vec<u8>) -> Weight {
+		0
+	}
+}
+
+/// Routines related to the upward message passing.
+impl<T: Trait> Module<T> {
+	pub(super) fn clean_ump_after_outgoing(outgoing_para: ParaId) {
+		<Self as Store>::RelayDispatchQueueSize::remove(&outgoing_para);
+		<Self as Store>::RelayDispatchQueues::remove(&outgoing_para);
+
+		// Remove the outgoing para from the `NeedsDispatch` list and from
+		// `NextDispatchRoundStartWith`.
+		//
+		// That's needed for maintaining invariant that `NextDispatchRoundStartWith` points to an
+		// existing item in `NeedsDispatch`.
+		<Self as Store>::NeedsDispatch::mutate(|v| {
+			if let Ok(i) = v.binary_search(&outgoing_para) {
+				v.remove(i);
+			}
+		});
+		<Self as Store>::NextDispatchRoundStartWith::mutate(|v| {
+			*v = v.filter(|p| *p == outgoing_para)
+		});
+	}
+
+	/// Check that all the upward messages sent by a candidate pass the acceptance criteria. Returns
+	/// false, if any of the messages doesn't pass.
+	pub(crate) fn check_upward_messages(
+		config: &HostConfiguration<T::BlockNumber>,
+		para: ParaId,
+		upward_messages: &[UpwardMessage],
+	) -> bool {
```

Comment:
The UX of this concerns me a bit: so if sending 100 messages fail at 42-nd message the operators of the parachain would have to manually try to understand which one caused the failure and why and so on

Reply:
I don't think we can do with the `bool` anything: down the line it will have to fit thru the `decl_error`'s bottleneck.

But at least we can log?

Reply:
I've created an issue https://github.com/paritytech/polkadot/issues/1898